### PR TITLE
feat: AM-API Exception Handling and Swagger Support

### PR DESCRIPTION
**Pull request by abhijitdiwan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/AM-306

### Change description ###
Feat: APIs Exception Handling and Swagger Support

### CODE REVIEW

**Review by shashank-rastogi — CHANGES_REQUESTED** [review]

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java [hunk 1]
```diff
@@ -16,6 +18,7 @@
 
 @Data
 @Builder
+@JsonDeserialize(builder = ExplicitAccessGrant.ExplicitAccessGrantBuilder.class)
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: You mean change description? if yes updated

File: src/main/java/uk/gov/hmcts/reform/amapi/controllers/AmLibProxyController.java [hunk 2]
```diff
@@ -23,28 +31,57 @@
...
     }
 
+    @ApiOperation(value = "Filter access to resource", response = ExplicitAccessGrant.class)
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = "Successfully returns filtered output"),
+        @ApiResponse(code = 415, message = "Unsupported media type.Excepted json"),
+        @ApiResponse(code = 401, message = "Accessing the resource you were trying to reach is forbidden."
+            + "Please use a valid access token"),
+        @ApiResponse(code = 400, message = "Incomplete request information or Malformed input request"),
+        @ApiResponse(code = 500, message = "Internal server error")
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/amapi/controllers/AmLibProxyController.java]: Tested this and I have not get 404 comes when url is like filter-resource. Only get 404 when I got invalid URL. (eg. /invalidUrl or /invalidFiterResource)

File: src/main/java/uk/gov/hmcts/reform/amapi/util/ErrorConstants.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+@SuppressWarnings({
+    "checkstyle:HideUtilityClassConstructor",
+    "PMD.FieldNamingConventions"
+})
+public class ErrorConstants {
+
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/amapi/util/ErrorConstants.java]: seems PMD ask for class Avoid constants in interfaces. Interfaces define types, constants are implementation details better placed in classes or enums. See Effective Java, item 19.

File: src/main/java/uk/gov/hmcts/reform/amapi/controllers/AmLibProxyController.java [hunk 4]
```diff
@@ -23,28 +31,57 @@
     @Autowired
     private AccessManagementService am;
 
+    @ApiOperation(value = "Grant resource access to user", response = ExplicitAccessGrant.class)
+    @ApiResponses(value = {
+        @ApiResponse(code = 201, message = "Successfully granted the access to resource for user"),
+        @ApiResponse(code = 415, message = "Unsupported media type.Excepted json"),
+        @ApiResponse(code = 401, message = "Accessing the resource you were trying to reach is forbidden."
+            + "Please use a valid access token"),
+        @ApiResponse(code = 400, message = "Incomplete request information or Malformed input request"),
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/amapi/controllers/AmLibProxyController.java]: updated

File: src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
+package uk.gov.hmcts.reform.amapi.controller;
+
+import com.fasterxml.jackson.core.JsonPointer;
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerExceptionAdviceTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        String invalidJson = "{\n"
+            + "  \"userId\": \"${accessorId}\",\n"
+            + "  \"userRoles\": [\n"
+            + "    \"caseworker\"\n"
+            + "  ],\n"
+            + "  \"resource\": {\n"
+            + "    \"id\": \"${resourceId}\",\n"
+            + "    \"definition\": {\n"
+            + "      \"serviceName\": \"cmc\",\n"
+            + "      \"resourceType\": \"case\",\n"
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerExceptionAdviceTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerExceptionAdviceTest.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
+package uk.gov.hmcts.reform.amapi.controller;
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/test/java/uk/gov/hmcts/reform/amapi/controller/ControllerExceptionAdviceTest.java]: done

**Review by shashank-rastogi — APPROVED** [review]

**Pull request merged**